index.ts

```diff
@@ -185,53 +186,62 @@ function suiteClassCallback(target: SuiteCtor, context: TestFunctions) {
-        try {
-          let method = <Function>prototype[key];
-          if (method === target) {
-            return;
-          }
-
-          let testName = method[testNameSymbol];
-          let shouldSkip = method[skipSymbol];
-          let shouldOnly = method[onlySymbol];
-          let shouldPending = method[pendingSumbol];
-
-          let testFunc = (shouldSkip && context.it.skip)
-                         || (shouldOnly && context.it.only)
-                         || context.it;
-
-          if (testName || shouldOnly || shouldPending || shouldSkip) {
-            testName = testName || (<any>method).name;
-            if (shouldPending && !shouldSkip && !shouldOnly) {
-              context.it.skip(testName);
-            } else if (method.length > 0) {
-              testFunc(testName, noname(function(this: Mocha.ITestCallbackContext, done) {
-                applyDecorators(this, prototype, method, instance);
-                applyTestTraits(this, instance, method);
-                return method.call(instance, done);
-              }));
-            } else {
-              testFunc(testName, noname(function(this: Mocha.ITestCallbackContext) {
-                applyDecorators(this, prototype, method, instance);
-                applyTestTraits(this, instance, method);
-                return method.call(instance);
-              }));
-            }
-          }
-        } catch (e) {
-          // console.log(e);
-        }
-      });
+		function runTest(prototype: any, method: Function) {
+			let testName = method[testNameSymbol];
+			let shouldSkip = method[skipSymbol];
+			let shouldOnly = method[onlySymbol];
+			let shouldPending = method[pendingSumbol];
+
+			let testFunc = (shouldSkip && context.it.skip)
+										 || (shouldOnly && context.it.only)
+										 || context.it;
+
+			if (testName || shouldOnly || shouldPending || shouldSkip) {
+				testName = testName || (<any>method).name;
+				if (shouldPending && !shouldSkip && !shouldOnly) {
+					context.it.skip(testName);
+				} else if (method.length > 0) {
+					testFunc(testName, noname(function(this: Mocha.ITestCallbackContext, done) {
+						applyDecorators(this, prototype, method, instance);
+						applyTestTraits(this, instance, method);
+						return method.call(instance, done);
+					}));
+				} else {
+					testFunc(testName, noname(function(this: Mocha.ITestCallbackContext) {
+						applyDecorators(this, prototype, method, instance);
+						applyTestTraits(this, instance, method);
+						return method.call(instance);
+					}));
+				}
+			}
     }
 
-    // run all tests along the inheritance chain
+    // collect all tests along the inheritance chain, allow overrides
+		const collectedTests: { [key: string]: Array<any> } = {};
     let currentPrototype = prototype;
     while (currentPrototype !== Object.prototype) {
-			runTests(currentPrototype);
+
+      Object.getOwnPropertyNames(currentPrototype).forEach(key => {
```

Comment:
I think `getProperyNames()` will enumerate the methods in the parent classes and you may skip the recursion that is walking up the prototype chain here. Could you give it a try?

Also that way if you override a method but don't add the `@test` keyword, you will effectively hide the test in the base class for this derived class.

Reply:
I will give that a try.

Reply:
Tried that. At the time that the tests are collected and run, the instance is not initialised and will be once the before each test hook was run. Besides of that, there is no getPropertyNames() method on neither Object nor target nor the prototype. And calling getOwnPropertyNames() on the instance will also return nothing at all.

Reply:
Object.keys(), but I tried it and it won't help. I am ok with the current implementation.